feat: add ChainController

ccharly opened the pull request:
## Explanation

This new controller will be responsible of exposing Snaps that implement the newly defined Chain API.

This controller also implements the chain API itself.

Reference: https://github.com/MetaMask/chain-api/

## Changelog

### `@metamask/chain-controller`

- **Added**: New `ChainController`: exposes/interacts with chain API providers 

## Checklist

- [x] I've updated the test suite for new or updated code as appropriate
- [x] I've updated documentation (JSDoc, Markdown, etc.) for new or updated code as appropriate
- [x] I've highlighted breaking changes using the "BREAKING" category above as appropriate

ccharly commented on packages/chain-controller/src/ChainController.ts:
```diff
@@ -0,0 +1,167 @@
+        ...state,
+      },
+    });
+
```
Good point, I did change now to use the `Messenger` now!

ccharly commented on packages/chain-controller/src/ChainController.ts:
```diff
@@ -0,0 +1,167 @@
+>;
+
+const defaultState: ChainControllerState = {
+  dummy: '',
```
I got rid of it, I thought a state was required, but after checking the type, I was able to make this work with an empty state.

mcmire commented on packages/chain-controller/src/SnapChainProviderClient.test.ts:
```diff
@@ -0,0 +1,74 @@
+
+  describe('getBalances', () => {
+    it('dispatch chain_getBalances', async () => {
+      const client = new SnapChainProviderClient(snapClient);
+
+      const address = 'bc1qrp0yzgkf8rawkuvdlhnjfj2fnjwm0m8727kgah';
+      const scope = 'bip122:000000000019d6689c085ae165831e93';
+      const asset = `${scope}/asset:0`;
+
+      const request = makeRequest({
+        method: ChainRpcMethod.GetBalances,
+        params: {
+          scope,
+          accounts: [address],
+          assets: [asset],
+        },
+      });
+
+      const response = {
+        balances: {
+          [address]: {
+            [asset]: {
+              amount: '70.02255139',
+            },
+          },
+        },
+      };
+
+      snapController.handleRequest.mockResolvedValue(response);
+      const result = await client.getBalances(scope, [address], [asset]);
+      expect(snapController.handleRequest).toHaveBeenCalledWith(request);
+      expect(result).toStrictEqual(response);
```
Nit: To help make it abundantly clear what exactly is being tested here, what do you think about using line breaks to divide the test into different "phases"? For instance, in this case you could remove the empty lines between all of the steps above this line, and then surround this line with empty lines to separate it visually. See here for more details: https://github.com/MetaMask/contributor-docs/blob/main/docs/unit-testing.md#-highlight-the-exercise-phase

``` suggestion
      const client = new SnapChainProviderClient(snapClient);
      const address = 'bc1qrp0yzgkf8rawkuvdlhnjfj2fnjwm0m8727kgah';
      const scope = 'bip122:000000000019d6689c085ae165831e93';
      const asset = `${scope}/asset:0`;
      const request = makeRequest({
        method: ChainRpcMethod.GetBalances,
        params: {
          scope,
          accounts: [address],
          assets: [asset],
        },
      });
      const response = {
        balances: {
          [address]: {
            [asset]: {
              amount: '70.02255139',
            },
          },
        },
      };
      snapController.handleRequest.mockResolvedValue(response);

      const result = await client.getBalances(scope, [address], [asset]);

      expect(snapController.handleRequest).toHaveBeenCalledWith(request);
      expect(result).toStrictEqual(response);
```

ccharly replied:
If this match our conventions, then yes why not!

mcmire commented on packages/chain-controller/src/SnapHandlerClient.ts:
```diff
@@ -0,0 +1,116 @@
+      origin: this.#origin,
+      handler: this.#handlerType,
+      request,
+    }) as Promise<Json>;
```
Is there a reason why a type assertion needs to be used? It would be nice to not use them.

ccharly replied:
Well, `this.#handler` will returns a `Promise<unknown>` and I wanted to restrict the type a bit more.

In anyway, given the "dynamic nature" of submitting requests to a Snap through the `handleRequest` handler, having an `unknown` makes sense.

We do have another type assertion here which will be necessary to match our API: https://github.com/MetaMask/core/blob/4ce91aa555a0a688f2f371ae5547738c645d2d13/packages/chain-controller/src/SnapChainProviderClient.ts#L30

So I can probably get rid of this one yes!

ccharly replied:
I did remove the type assertion as we are doing this in the "API client" (see above)

mcmire reviewed (commented):
Hello! I am not sure whether more work was planned for this PR, but I made some suggestions based on how we've written controllers and tests for controllers in other places. One question I have is whether ChainController needs to be a controller? I didn't notice any state being set, but perhaps that was just left out for now.

ccharly commented on packages/chain-controller/src/ChainController.test.ts:
```diff
@@ -0,0 +1,133 @@
+import { ChainController } from './ChainController';
+
+describe('ChainController', () => {
+  const snapId = 'local:localhost:3000' as SnapId;
```
Thanks @mcmire for this! The main reason why I do use "shared variables" is to be more DRY.

But it's true that some of those variables could be mutated "behind the scene" resulting in a non-consistent state in some tests.

I'm definitely ok to adapt my tests to match our standard, so lemme refactor this! :)

ccharly commented on packages/chain-controller/src/ChainController.test.ts:
```diff
@@ -0,0 +1,133 @@
+    },
+  } as unknown as SnapController;
+
+  const messenger = {
```
Yup, good point. Following your first advices (to match our standard way of doing this), I ended up removing every mentions of the `SnapController`, focusing solely on mock implementations, which makes the overall tests a bit more "straightforward".

Thanks you for this :)

mcmire replied:
Thanks, the changes look better!

ccharly commented on packages/chain-controller/src/ChainController.test.ts:
```diff
@@ -0,0 +1,133 @@
+
+    it('fails to register another provider for an existing scope', () => {
+      const controller = makeController();
+      const anotherSnapId = 'local:localhost:4000' as SnapId;
```
I must say, that I don't know... Since it's an opaque type (and I haven't seen any "factory"/"builder" methods to make one out of a `string`, I do use type assertion like here: https://github.com/MetaMask/snaps/blob/main/packages/snaps-controllers/src/snaps/SnapController.test.tsx#L5528-L5529

mcmire replied:
Ah, I see, interesting. Perhaps this is the only way, but I've asked the Snaps team about this just in case there is a better way.

ccharly commented on packages/chain-controller/src/ChainController.test.ts:
```diff
@@ -0,0 +1,133 @@
+    it('returns false if there is no known provider', () => {
+      const controller = makeController();
+
+      expect(controller.hasProviderFor(scope)).toBe(false);
```
I guess this makes sense, lemme move this to another `describe`.

ccharly commented on packages/chain-controller/src/ChainController.ts:
```diff
@@ -0,0 +1,160 @@
+
+const controllerName = 'ChainController';
+
+export type ChainControllerState = Record<string, Json>;
```
Two things:

1. Good catch, I started using `Record` to make the type system happy, my initial ideal was to use `{}` which seems to be forbidden

2. For now, no, we have no "persisted" state, we do have a "in-memory" one though, for the providers. So ultimately, this could be a plain class as-is, yes. In the future we might need a persistent state yes.

So what do you recommend/prefer here?
- I can make it a "look-alike" controller (like the messenger controller), and we could derive from the `BaseController` later on only if we need a state?
- Or I keep it as is (given that we might have a persistent state "soon")?

ccharly replied:
Also, if we keep it as is, I could go for `Record<string, never>` instead, so this way this is similar to `{}` (since no other construction will match this type). WDYT?

mcmire replied:
Okay, good questions. I guess if we plan on having persisted state in the future, we can keep it as a controller and use `Record<string, never>` for the state.

ccharly commented on packages/chain-controller/src/ChainController.ts:
```diff
@@ -0,0 +1,160 @@
+    throw new Error(error);
+  }
+
+  getBalances = async (
```
We need a provider for a given scope, so yes they need to be bound to this instance.

ccharly commented on packages/chain-controller/src/ChainController.ts:
```diff
@@ -0,0 +1,160 @@
+    snapId: SnapId,
+  ): SnapChainProviderClient {
+    // TODO: Should this be idempotent?
+    const client = this.#snapClient.withSnapId(snapId as SnapId);
```
This one is extraneous indeed, I'm surprised our static analyser (or at least the linter), did not spot that one, I'll remove it! Thanks :)

mcmire commented on packages/chain-controller/src/SnapChainProviderClient.test.ts:
```diff
@@ -0,0 +1,72 @@
+  const snapController = {
+    handleRequest,
+  };
+  const snapClient = new SnapHandlerClient({
```
Similar to other comment, what are your thoughts on creating a new SnapHandlerClient for each test that needs one?

mcmire commented on packages/chain-controller/src/SnapHandlerClient.test.ts:
```diff
@@ -0,0 +1,72 @@
+import { SnapHandlerClient } from './SnapHandlerClient';
+
+describe('SnapHandlerClient', () => {
+  const handleRequest = jest.fn();
```
What are your thoughts on recreating this for each test? Otherwise the state of this mock function may bleed into other tests.

mcmire commented on packages/chain-controller/src/SnapHandlerClient.test.ts:
```diff
@@ -0,0 +1,72 @@
+  const handleRequest = jest.fn();
+
+  const snapId = 'local:localhost:3000' as SnapId;
+  const snapController = {
```
Similar as previous — do you need the concept of a SnapController here? Maybe you just need to pass the `handleRequest` around.

mcmire reviewed (commented):
A few more comments, related to other ones I've made.

Also, the README lists the packages in the monorepo and provides a diagram. We don't update this automatically right now — would you mind running `yarn update-readme-content`? That command should add `chain-controller` to the list/diagram.

wachunei commented on packages/chain-controller/src/ChainController.test.ts:
```diff
@@ -0,0 +1,152 @@
+const snapId = 'local:localhost:3000' as SnapId;
+
+const address = 'bc1qrp0yzgkf8rawkuvdlhnjfj2fnjwm0m8727kgah';
+const scope = 'bip122:000000000019d6689c085ae165831e93';
+const asset = `${scope}/asset:0`;
```
I noticed [CAIP-2](https://chainagnostic.org/CAIPs/caip-2) has been adopted for the chain identifier. The asset variable seems to be adopting a similar definition of [CAIP-19](https://chainagnostic.org/CAIPs/caip-19), but the asset namespace was set to `asset`. [CAIP-20](https://chainagnostic.org/CAIPs/caip-20) specifies the a chain agnostic asset namespace we could use. 

We are currently discussing the adoption of these CAIPs in [this slack thread](https://consensys.slack.com/archives/C06EULX1456/p1715185176384949) and it would be very valuable to hear the rationale of the `asset` asset namespace decision or other thoughts about adopting these 🙏

ccharly replied:
Ah yes sorry, I used `asset` here, cause it's "accepted" as a valid asset namespace. And since this is a test with a mock, the actual value is not really important (as long as it's a valid CAIP-19 identifier at least).

I will change that to use `slip44` here, so that it would also "match" the actual `scope`.

Thanks for spotting this :)

ccharly commented:
@mcmire I did update the README, but I had to hack the script locally because it was removing the `accounts-controller` from the README because of the `.slice(1)` [here](https://github.com/MetaMask/core/blob/main/scripts/update-readme-content.ts#L62).

I didn't want to fix this in this PR, cause I'm not sure it's still required or not.

ccharly commented:
I've added the `DO-NOT-MERGE` label, cause I've added `skipLibCheck=true` for tsbuild temporarily, otherwise the CI complains (something the `jsx-runtime`, we faced that in another PR and that might be related to some Snaps dependencies, I'll wait for more info on this before removing it from the PR)

gantunesr commented on packages/chain-controller/src/SnapHandlerClient.ts:
```diff
@@ -0,0 +1,130 @@
+   */
+  constructor({
+    handler,
+    // Follow same pattern than for @metamask/keyring-api
+    snapId = 'undefined' as SnapId,
```
What is the purpose of this?

ccharly replied:
This is also coming from the `keyring-api`, the rationale being that you can construct a client without providing any `snapId` and you can then constraint it using `withSnapId` method.

Also, if we were to use an optional here, that would require us to "over-check" if `.snapId` is defined or not, which will result in a more complex code logic IMO.

Anyway, if you submit a request with an invalid Snap id, the request will fail, so we're cover here (not sure if that's the best pattern though...).

As a side note, I just re-used the same pattern than for the `keyring-api` too 😅

ccharly replied:
That being said, after re-reading the logic, I guess I could get rid of this pattern...

Lemme try to address that, cause I feel like this is gonna make things a bit simpler and straightforward!

gantunesr replied:
> _if we were to use an optional here, that would require us to "over-check" if .snapId_

I think I don't follow what you mean here. Is this just to avoid the TS error because of an optional property? Because setting it to `undefined` will be the same as having it as optional in terms of execution

gantunesr commented on packages/chain-controller/src/SnapHandlerClient.test.ts:
```diff
@@ -0,0 +1,68 @@
+      success: true,
+    };
+
+    it('should call a method and return the result', async () => {
```
Following our testing practices, we shouldn't start a test name with `should`.

https://github.com/MetaMask/contributor-docs/blob/main/docs/e2e-testing.md#test-names

```suggestion
    it('returns a result when a method is called', async () => {
```

gantunesr commented on packages/chain-controller/README.md:
```diff
@@ -0,0 +1,15 @@
+# `@metamask/chain-controller`
```
I see that there are a lot of parts to this controller package with the existence of `ChainController`, `SnapHandlerSender`, `SnapChainProviderClient`, etc.

It would be good to have a diagram and/or short explanation in the README about how the relationship between each of these components.

ccharly replied:
That's a good idea, I'll see what I can do! Thanks

ccharly commented on packages/chain-controller/src/ChainController.ts:
```diff
@@ -0,0 +1,197 @@
+
+const controllerName = 'ChainController';
+
+export type ChainControllerState = Record<string, never>;
+
+export type ChainControllerGetStateAction = ControllerGetStateAction<
+  typeof controllerName,
+  ChainControllerState
+>;
+
+export type AllowedActions = SnapControllerHandleSnapRequestAction;
+
+export type ChainControllerActions = never;
+
+export type ChainControllerChangeEvent = ControllerStateChangeEvent<
+  typeof controllerName,
+  ChainControllerState
+>;
+
+export type AllowedEvents = ChainControllerEvents;
+
+export type ChainControllerEvents = ChainControllerChangeEvent;
+
+export type ChainControllerMessenger = RestrictedControllerMessenger<
+  typeof controllerName,
+  ChainControllerActions | AllowedActions,
+  ChainControllerEvents | AllowedEvents,
+  AllowedActions['type'],
+  AllowedEvents['type']
+>;
```
I feel like those needs to live here, I just followed the same patterns than some other controllers like:
- https://github.com/MetaMask/core/blob/main/packages/user-operation-controller/src/UserOperationController.ts#L102-L118
- https://github.com/MetaMask/core/blob/main/packages/accounts-controller/src/AccountsController.ts#L35-L119
- https://github.com/MetaMask/core/blob/main/packages/keyring-controller/src/KeyringController.ts#L84-L215

gantunesr replied:
Yeah, it's an established pattern in our controller files but in general I think having types in a separate file is a better practice. At the end of opinion it's just an opinion, so this way is ok too.

ccharly commented on packages/chain-controller/src/SnapHandlerClient.ts:
```diff
@@ -0,0 +1,130 @@
+   */
+  constructor(
+    handler: Handler,
+    handlerType: HandlerType,
```
I guess so yes. Right now, it's not yet available on Snaps side, and since this class is mostly generic I just wanted to use "any handler".

This is highly inspired from the `keyring-api` clients like this one:
- https://github.com/MetaMask/keyring-api/blob/main/src/KeyringSnapRpcClient.ts

ccharly commented:
Removing the `DO-NOT-MERGE` label for now, to have a successful CI and allow preview builds

ccharly commented:
@metamaskbot publish-preview

mcmire commented:
@ccharly Thank you for the changes! I agree it looks a lot better. Nothing jumps out at me, so feel free to get approval from your team and continue on with this PR :)

gantunesr commented on packages/chain-controller/src/SnapChainProviderClient.test.ts:
```diff
@@ -0,0 +1,79 @@
+/**
+ * Builds a Snap chain API request.
+ *
+ * @param options0 - Chain API request object.
+ * @param options0.method - Chain API method to be called.
+ * @param options0.params - Chain API parameters.
```
Is this a typo? `options0`?

ccharly replied:
ah that was a suggestion my editor I guess, because I didn't give a name to the object I guess..

Do you want me to change that?

ccharly replied:
I went for `request` instead, thanks!

gantunesr reviewed (dismissed):
LGTM. Just one minor comment
